soc: xilinx: zynq7000: add pinctrl driver + digilent zybo board

henrikbrixandersen opened the pull request:
- Add pinctrl driver for the Xilinx Zynq-7000 series SoC.
- Add pinctrl support to the Xilinx Zynq-7000 PS UART driver.
- Add board definition for the Digilent Zybo to get code coverage for the newly added pinctrl driver.

henrikbrixandersen committed:
scripts: dts: gen_defines: define _TOKEN + _UPPER_TOKEN for string arrays

Generate definitions for *_IDX_n_TOKEN and *_IDX_n_UPPER_TOKEN for all
string-array type properties.

Signed-off-by: Henrik Brix Andersen <henrik@brixandersen.dk>

henrikbrixandersen committed:
dts: bindings: pinctrl: add Xilinx Zynq-7000 pinctrl devicetree binding

Add devicetree binding for the Xilinx Zynq-7000 series pinctrl.

Signed-off-by: Henrik Brix Andersen <henrik@brixandersen.dk>

henrikbrixandersen committed:
soc: arm: xilinx: zynq7000: add pinctrl header file

Add Xilinx Zynq-70000 pinctrl header file to define SoC specific
pinctrl_soc_t structure. This is used to store pin configurations for the
pinctrl driver.

Signed-off-by: Henrik Brix Andersen <henrik@brixandersen.dk>

henrikbrixandersen committed:
dts: arm: xilinx: zynq7000: add pinctrl node

Add devicetree node for the pinctrl of the Xilinx Zynq-7000 series. Pinctrl
is set through a subset of the System Level Control Registers (SLCR), which
is accessed through the syscon driver.

Signed-off-by: Henrik Brix Andersen <henrik@brixandersen.dk>

henrikbrixandersen committed:
drivers: pinctrl: add pinctrl driver for the Xilinx Zynq-7000

Add pinctrl driver for the Xilinx Zynq-7000 series SoCs.

Signed-off-by: Henrik Brix Andersen <henrik@brixandersen.dk>

henrikbrixandersen committed:
drivers: serial: xilinx: change prompt to reflect supported SoCs

Change the Kconfig prompt for the Xilinx Zynq-7000/ZynqMP PS UART driver to
indicate whichs SoCs it supports.

Signed-off-by: Henrik Brix Andersen <henrik@brixandersen.dk>

henrikbrixandersen committed:
drivers: serial: xilinx: uartps: add pinctrl support

Add pinctrl support to the Xilinx Zynq-7000/ZynqMP PS UART driver.

Signed-off-by: Henrik Brix Andersen <henrik@brixandersen.dk>

henrikbrixandersen committed:
boards: arm: zybo: add board definition for the Digilent Zybo

Add board definition for the Xilinx Zynq-7000-based Digilent Zybo.

Signed-off-by: Henrik Brix Andersen <henrik@brixandersen.dk>

henrikbrixandersen committed:
soc: arm: xilinx: zynq7000: unlock the slcr at boot

Unlock the Xilinx Zynq-7000 System Level Control Registers (SLCR) at boot
to allow write access using the generic syscon driver.

Since the generic syscon driver uses the DEVICE_MMIO_* APIs for
memory-mapped IO we can remove the dedicated SLCR MMU region.

Signed-off-by: Henrik Brix Andersen <henrik@brixandersen.dk>

ibirnbaum commented:
Hi Henrik,

just a few quick initial comments after just having received the notification and before leaving for the rest of the day:
- Regarding the altered SLCR mapping in the SoC code: do you intend to permanently leave the SLCR unlocked, or did I miss on-demand locking in the pinctrl driver? As far as I remember, Xilinx encourages re-locking it once a write operation on the SLCR registers has been performed. Please note that the GEM driver accesses the SLCR in order to change the clock dividers upon link speed chage reported by the PHY, and once this has been done, the registers are re-locked. This already happens in the initial setup of the driver, so once the GEM is initialized, the registers will be locked.
- Similarly to what I wrote in the CAN driver PR regarding a clock controller DT node or driver the other day, I don't really get the point of a pin controller driver for the Zynq or UltraScale. From my point of view, all the I/O and clock configuration (disregard special scenarios like the afforementioned GEM link speed change) comes from Vivado rather than the Zephyr config and thanks to the boot mechanism including the bitstream programming, Zynqs and Ultrascales can be considered fully set up with regards to I/O and clocks by the time Zepyhr boots - all that has been handled by the FSBL. Why should we meddle with the MIO pin assignments from within Zephyr afterwards?

henrikbrixandersen commented:
> * Regarding the altered SLCR mapping in the SoC code: do you intend to permanently leave the SLCR unlocked, or did I miss on-demand locking in the pinctrl driver? As far as I remember, Xilinx encourages re-locking it once a write operation on the SLCR registers has been performed. Please note that the GEM driver accesses the SLCR in order to change the clock dividers upon link speed chage reported by the PHY, and once this has been done, the registers are re-locked. This already happens in the initial setup of the driver, so once the GEM is initialized, the registers will be locked.

Yes. The intent is to leave the SLCR permanently unlocked. I know that the registers allows for it to be locked again, but I have not seen anywhere that this is a recommended by Xilinx? Do you have a reference to this? What would the benefit be? Any code, that can write to the SLCRs, can just unlock them?

Leaving the SLCRs unlocked has a couple of benefits: 
- No need for locking (by semaphore or mutex) access to the SLCRs (if we use unlock -> modify -> lock, we need to guard against the thread having unlocked the SLCRs being pre-empted by another thread also unlocking-modifying-locking the SLCRs, causing the SLCRs to be locked when the first thread attempts to write to them). This scheme would require a custom syscon driver for the Zynq-7000 SLRCs for very little gain.
- No locking means we can use the generic syscon driver for accessing the SLCRs

The Linux kernel also unlocks the Zynq-7000 SLCRs at boot time and leaves the unlocked: https://elixir.bootlin.com/linux/latest/source/arch/arm/mach-zynq/slcr.c#L220

I was not aware of the GEM driver accessing the SLCRs directly. This needs to be updated (but by using a clock control driver, not by accessing the SLCRs directly, see my next paragraph).

> * Similarly to what I wrote in the CAN driver PR regarding a clock controller DT node or driver the other day, I don't really get the point of a pin controller driver for the Zynq or UltraScale. From my point of view, all the I/O and clock configuration (disregard special scenarios like the afforementioned GEM link speed change) comes from Vivado rather than the Zephyr config and thanks to the boot mechanism including the bitstream programming, Zynqs and Ultrascales can be considered fully set up with regards to I/O and clocks by the time Zepyhr boots - all that has been handled by the FSBL. Why should we meddle with the MIO pin assignments from within Zephyr afterwards?

If your use-case is what you describe above, you probably do not need to alter the pinctrl settings from Zephyr. In that case, you can just leave `CONFIG_PINCTRL=n` and no Zephyr code will touch the pinctrl settings.

We have a different use-case. We have a modular hardware design:
- The CPU board containing the Zynq-7000 is generic and programmed at the factory with a generic, minimalistic FSBL + U-Boot in QSPI.
- At boot time, the FSBL applies a minimal configuration and passes control to U-Boot.
- U-Boot probes the hardware and determines the type of base board(s) the CPU board is connected to, loads and boots the corresponding OS + software configuration from external media.
- The OS then configures clocking, sets up pinctrl, enables peripherals, loads the bitstream corresponding to the hardware, etc.

For this use-case we need both pinctrl and clock control. Pinctrl is introduced here, next up is a proper clock control driver. The Linux kernel includes both a pinctrl and a clock control driver for the Zynq-7000 to meet these needs.

Other use-cases for pinctrl + clock control:
- When using Zephyr on a generic Zynq-7000 development board using the standard U-Boot configuration for that board, a pinctrl + clock control driver would allow for the OS the configure the development board using shield definitions/devicetree overlays to accommodate to the various attached Pmods, shields, and other extensions.
- Since the FSBL + U-Boot is often programmed into e.g. QSPI flash at the factory, they can be difficult to update over-the-air (at least without risking bricking the device due to a failed update). Having full pinctrl + clock control support within the OS allows "patching" wrong settings in the deployed FSBL from an OS update received OTA.

gmarull commented on dts/bindings/pinctrl/xlnx,pinctrl-zynq.yaml:
```diff
@@ -0,0 +1,190 @@
+      power-source:
+        enum: [1, 2, 3, 4]
+        description: |
+          IO buffer type. Sets the IO_Type field in the MIO_PIN_xx SLCR register.
+
+          1 or IO_STANDARD_LVCMOS18
+          2 or IO_STANDARD_LVCMOS25
+          3 or IO_STANDARD_LVCMOS33
+          4 or IO_STANDARD_HSTL
+
+      slew-rate:
+        enum: [0, 1]
+        description: |
+          IO buffer edge rate. Applicable when the power-source (IO_type) is LVCMOS18, LVCMOS25, or
+          LVCMOS33. Sets the Speed field in the MIO_PIN_xx SLCR register.
+
+          0 or IO_SPEED_SLOW
+          1 or IO_SPEED_FAST
```
nit: mention definitions are provided in `pinctrl-zynq.h`

gmarull commented on boards/arm/zybo/zybo.dts:
```diff
@@ -0,0 +1,86 @@
+		sw1 = &btn5;
+	};
+
+	cpus {
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		cpu0: cpu@0 {
+			device_type = "cpu";
+			compatible = "arm,cortex-a9";
+			reg = <0>;
+		};
+
+		cpu1: cpu@1 {
+			device_type = "cpu";
+			compatible = "arm,cortex-a9";
+			reg = <1>;
+		};
+	};
+
+	sram0: memory@0 {
+		compatible = "mmio-sram";
+		reg = <0x00000000 DT_SIZE_M(512)>;
+	};
```
question: why is this defined at board level? is it because of the nature of the SoC (FPGA)?

ibirnbaum replied:
> question: why is this defined at board level? is it because of the nature of the SoC (FPGA)?

Regarding the RAM: the Zynq has an integrated DDR memory controller, but other than some on-chip memory, no internal RAM. So the RAM size depends on what is actually soldered onto the respective board. The same applies to flash memory.

Regarding the CPUs: the Zynq is available in two general flavors: XC7Zxxx, which has two A9 cores, and XC7ZxxxS, which only has one core. Several sub-models, which differ in I/O pin count and FPGA type / logic cell count (Artix-7 for the smaller models, Kintex-7 for the larger models), exist for both of these. However, despite a varying set of actually available peripherals (due to I/O pin constraints) and programmable logic space, all those chips always have the same set of on-board peripherals. Therefore, the CPU nodes were moved to the board level, initially for qemu_cortex_a9, in order to:
- not to have two dts files with what may eventually become a sizeable set of supported peripherals, whose only difference is the 1 vs. 2 CPU core count,
- allow selective SMP / single-core operation with the same board by means of overlays or a similar solution (comp. qemu_cortex_a53 which optionally has an _smp variant).
Another minor factor for this move was the consistency between the device tree used for building Zephyr and the device tree provided to QEMU for the qemu_cortex_a9 target. The latter effectively specifies a single-core Zynq, as QEMU is acting up if a 2nd core is present but not taken out of the boot loop by the OS due to a lack of SMP capability.
All this was introduced in #40964.

henrikbrixandersen replied:
As @ibirnbaum wrote, the SDRAM is external and thus board-specific. I would like to move the CPU nodes into the SoC DTS file (we can easily handle single- vs. dual-core in those, even from overlays) along with some other clean-ups in a later PR.

gmarull reviewed (dismissed):
lgtm, great job on keeping compatibility with upstream bindings

ibirnbaum commented:
> Yes. The intent is to leave the SLCR permanently unlocked. I know that the registers allows for it to be locked again, but I have not seen anywhere that this is a recommended by Xilinx? Do you have a reference to this?

Admittedly, it's not stated explicitly in the TRMs, for me, an obvious clue was that the PS7Init code re-locks the registers after each initialization step. The closest to an official statement regarding the issue is probably Xilinx's Answer Record 47570, which refers to an early version of the Boot ROM not re-locking the registers before exiting. The record states: "Leaving the SLCR registers unlocked exposes them to being accidentally overwritten". Although protection of the registers against rogue write accesses is the obvious purpose, it's admittedly more likely that those will try to write to NULL rather than somewhere beyond 0xF8000000. If leaving the unlocked helps against the need for an extra mutex, fine, let's go for that.

> I was not aware of the GEM driver accessing the SLCRs directly. This needs to be updated (but by using a clock control driver, not by accessing the SLCRs directly, see my next paragraph).

That sounds like a good idea for unified and abstracted clock configuration access, and most of the peripherals contained in the Zynq/UltraScale are simple enough that a one-time config will do, but if you're already working on that driver or are about to start working on it, could you please consider the following for the GEM(s):

- The clock controller driver should either have a means of obtaining the frequency of the PLL that drives the respective GEM and setting the pre-scalers without needing any knowledge about what the clock source is (effectively, allowing the modification of DIVISOR/DIVISOR1 in GEMx_CLK_CTRL without having to set SRCSEL and CLKACT as well) or something like a set_clock_frequency(hz) function which converts a frequency value into DIVISOR/DIVISOR1 values. This is required for TX clock re-configuration upon link speed change.
- The driver or rather its DT integration should allow the specification of EMIO clock sources independent of the ususal ARM/DDR/IOPLL -> DIVISOR -> DIVISOR1 structure so that GEMx_CLK_CTRL.SRCSEL can be set to EMIO and DIVISOR/DIVISOR1 re-calculation still works. This of course applies to any peripheral which can be clocked via EMIO.
- For the GEM, clocking information should be optional. When using the GEM in conjunction with the GMII-to-RGMII IP Core, clock frequency information is neither available nor required. Still, the clock source must be configurable to EMIO.

Also, there should be consideration for the fact that a Zynq/UltraScale based system may contain IP Core based peripherals whose clocking and I/O configurations are completely outside the scope of Zephyr, such as the AXI GPIO controller IP core. Therefore, with regards to this example, requiring pin control and/or clock control information for *any* GPIO controller might be too broad.

> The OS then configures clocking, sets up pinctrl, enables peripherals, loads the bitstream corresponding to the hardware, etc.

Do you already have a bitstream loader module for Zephyr? I wrote one in C many years ago, IIRC that was for a Virtex FPGA. In case the bitstream interface still looks similar in the Zynq (block-wise loading, plus a few flags like done, CRC error etc.), I could have a look at whether this code can be ported or not in case you don't have a loader module yet.

henrikbrixandersen commented:
The `tests/kernel/mp/kernel.multiprocessing` test case is failing with an undefined reference to `arch_start_cpu`. As far as I know, Zephyr does not yet support SMP on aarch32, but yet `soc/arm/xilinx_zynq7000/xc7zxxx/Kconfig.defconfig.series` sets `CONFIG_MP_NUM_CPUS=2` (causing this test to be selected).

@microbuilder What would be the correct solution here? Set `CONFIG_MP_NUM_CPUS=1` for the dual-core Zynq-7000 Cortex-A9 for now even though it has two cores?

henrikbrixandersen committed:
drivers: ethernet: xlnx: gem: do not lock/unlock the SLCRs

Do not lock/unlock the System Level Control Registers (SLCRs) in the Xilinx
GEM ethernet driver. The SLRCs are unlocked once at boot time.

Signed-off-by: Henrik Brix Andersen <henrik@brixandersen.dk>

microbuilder commented:
> @microbuilder What would be the correct solution here? Set `CONFIG_MP_NUM_CPUS=1` for the dual-core Zynq-7000 Cortex-A9 for now even though it has two cores?

I need to look at the Zynq-7000 in a bit more detail, but that would probably be my suggestion as an interim solution.

henrikbrixandersen committed:
soc: arm: xilinx: zynq7000: default to 1 CPU core

Default to 1 CPU core on the Xilinx Zynq-7000 SoC series since Zephyr does
not yet suppport SMP on aarch32.

Signed-off-by: Henrik Brix Andersen <henrik@brixandersen.dk>

henrikbrixandersen commented:
> > @microbuilder What would be the correct solution here? Set `CONFIG_MP_NUM_CPUS=1` for the dual-core Zynq-7000 Cortex-A9 for now even though it has two cores?
> 
> I need to look at the Zynq-7000 in a bit more detail, but that would probably be my suggestion as an interim solution.

@microbuilder Thanks. I have added a commit setting `CONFIG_MP_NUM_CPUS=1` until we have aarch32 SMP support in the tree.

henrikbrixandersen commented:
The failing compliance check is due to the added `pinctrl_soc_pin_t` typedef.
